File: ui/types/response/prompts.ts
import type { ApiLogReponse } from "./logs"

import type { Tool } from './tools'

export interface Prompt {
  id: number
  system: string
  user: string | null
  key: string
  model: string
  model_id: number
  provider: string
  max_tokens: number
  temperature: number
  json_mode: boolean
  json_schema: string | null
  prompt_type: string
  is_chat: boolean
  version_id: number
  version_number: number
  system_version_diff: string | null
  user_version_diff: string | null
  updated_at: string
  tools: Tool[]
  supports_json: boolean
  supports_json_schema: boolean
  supports_tools: boolean
  is_reasoning: boolean
  reasoning_effort: string | null
}


// Message type for chat
export interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string;
  tool_calls?: ToolCall[];
  tool_call_id?: string; // For tool responses
  // Raw data handling - not sent to server but used for UI
  rawData?: string;
  showRawData?: boolean;
}

// Tool call structure
export interface ToolCall {
  // A unique identifier for the tool call, optional in streaming responses
  id?: string;
  // The index of the tool call in the list of tool calls
  index?: number;
  // The type of call. When streaming, the first chunk only will contain "function"
  type?: string;
  // The details of the function call
  function: FunctionCall;
}

export interface FunctionCall {
  // The name of the function to call, can be optional in streaming chunks
  name?: string;
  // A JSON string representing the arguments for the function call
  arguments: string;
}

// PROMPT EXECUTION RESPONSE
export interface PromptExecutionResponse {
  content: string;
  log: ApiLogReponse;
}

// OpenAI compatible API types
export interface ApiCompletionResponse {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: ApiChoice[];
  usage: ApiUsage;
}

export interface ApiChoice {
  index: number;
  message: Message;
  finish_reason: string;
  tool_calls?: ToolCall[];
}

export interface ApiUsage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
}

// For streaming responses
export interface ApiCompletionChunk {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: ApiChunkChoice[];
}

export interface ApiChunkChoice {
  index: number;
  delta: ApiDelta;
  finish_reason?: string;
}

export interface ApiDelta {
  content?: string;
  role?: string;
  tool_calls?: ToolCall[];
}


// PROMPT PERFORMANCE
export interface PromptEvalVersionPerformanceResponse {
    version_id: number,
    version_number: number,
    version_date: string,
    avg_score: number | null,
    run_count: number,
}

File: backend/src/controllers/types/response/prompts.rs
use serde::Serialize;

use crate::db::types::prompt::PromptRowWithModel;
use crate::controllers::types::response::tools::ToolResponse;


// GET PROMPT RESPONSE
#[derive(Debug, Serialize)]
pub struct PromptResponse {
    pub id: i64,
    pub key: String,
    pub system: String,
    pub user: Option<String>,
    pub model: String,
    pub model_id: i64,
    pub provider: String,
    pub max_tokens: i64,
    pub temperature: f64,
    pub json_mode: bool,
    pub json_schema: Option<String>,
    pub prompt_type: String,
    pub is_chat: bool,
    pub version_id: i64,
    pub version_number: i64,
    pub system_version_diff: Option<String>,
    pub user_version_diff: Option<String>,
    pub updated_at: String,
    pub tools: Vec<ToolResponse>,
    pub supports_json: bool,
    pub supports_json_schema: bool,
    pub supports_tools: bool,
    pub is_reasoning: bool,
    pub reasoning_effort: Option<String>
}


impl From<PromptRowWithModel> for PromptResponse {
    fn from(prompt: PromptRowWithModel) -> Self {
        PromptResponse {
            id: prompt.id,
            key: prompt.key,
            system: prompt.system,
            user: prompt.user,
            model: prompt.model_name.into(),
            model_id: prompt.model_id,
            provider: prompt.provider_name.into(),
            max_tokens: prompt.max_tokens,
            temperature: prompt.temperature,
            json_mode: prompt.json_mode,
            json_schema: prompt.json_schema,
            prompt_type: prompt.prompt_type,
            is_chat: prompt.is_chat,
            version_id: prompt.version_id,
            version_number: prompt.version_number,
            system_version_diff: prompt.system_diff,
            user_version_diff: prompt.user_diff,
            updated_at: prompt.updated_at.to_string(),
            tools: Vec::new(), // Initialize with empty vector, will be filled separately
            supports_json: prompt.supports_json,
            supports_json_schema: prompt.supports_json_schema,
            supports_tools: prompt.supports_tools,
            is_reasoning: prompt.is_reasoning,
            reasoning_effort: prompt.reasoning_effort
        }
    }
}
